python2/httplib2/__init__.py

```diff
@@ -1740,9 +1756,106 @@ def clear_credentials(self):
         self.credentials.clear()
         self.authorizations = []
 
+    def add_cookie(self, name, value, max_age=None, expires=None):
+        """Set new cookie or update the existing cookie value.
+
+        max_age:
+            Number of seconds until the cookie expires
+        expires:
+            The maximum lifetime of the cookie as an HTTP-date timestamp.
+            Format:
+                <day-name>, <day> <month> <year> <hour>:<minute>:<second> GMT
+            Example: Wed, 21 Oct 2015 07:28:00 GMT
+
+        max_age has higher priority than expires.
+        """
+        if max_age:
+            expire_time = time.time() + int(max_age)
+        elif expires:
+            expire_time = calendar.timegm(email.Utils.parsedate_tz(expires))
+        else:
+            expire_time = None
+
+        self.cookies[name] = (value, expire_time)
+
+    def add_http_cookie(self, http_cookie):
+        """Set cookie with HTTP-style header(Set-Cookie:) format string.
+
+        http_cookie example:
+            sid=vBj3zbrmgynsZ8ggVpmSIOc9bXAueT3z; HttpOnly;
+
+        It may contain multiple cookies separated by comma(RFC 2109).
+
+        Example:
+            BD_NOT_HTTPS=1; path=/; Max-Age=300, BIDUPSID=D2504FE922735DE;
+            expires=Thu, 31-Dec-37 23:55:55 GMT; max-age=2147483647; path=/;
+            domain=.example.com, PSTM=1595329370
+
+        The 'expires' also contains comma. Be careful!
+        """
+        if not http_cookie:
+            return
+
+        for cookie_str in re.split(
+                r'(?:,)\s*(?=[^;, ]+?=[^;, ]+)', http_cookie):
```

Comment:
Pre-compile is mandatory. You can save compilation time only if cookie storage is enabled.
```
class Http:
  def __init__(cookies=None):
    if cookies is not None:
      self.cookie_regex = re.compile(...)
```

Reply:
Moved to the CookieJar as a class variable.